File: app/src/main/java/com/savvasdalkitsis/gameframe/ip/repository/PreferenceIpRepository.java
package com.savvasdalkitsis.gameframe.ip.repository;

import com.savvasdalkitsis.gameframe.infra.preferences.RxSharedPreferences;
import com.savvasdalkitsis.gameframe.ip.model.IpAddress;
import com.savvasdalkitsis.gameframe.ip.model.IpBaseHostMissingException;

import rx.Observable;

public class PreferenceIpRepository implements IpRepository {

    private static final String PREF_IP_BASE = "pref_ip_base";
    private RxSharedPreferences preferences;

    public PreferenceIpRepository(RxSharedPreferences preferences) {
        this.preferences = preferences;
    }

    @Override
    public Observable<IpAddress> getIpAddress() {
        return preferences.getString(PREF_IP_BASE)
                .map(IpAddress::parse)
                .flatMap(ipAddress -> {
                    if (ipAddress.isValid()) {
                        return Observable.just(ipAddress);
                    } else {
                        return Observable.error(new IllegalStateException("saved ip address was not valid: " + ipAddress.toString()));
                    }
                })
                .onErrorResumeNext(error -> Observable.error(new IpBaseHostMissingException("Error trying to read ip address from repository", error)));
    }

    @Override
    public void saveIpAddress(IpAddress ipAddress) {
        preferences.setString(PREF_IP_BASE, ipAddress.toString());
    }

}

File: app/src/main/java/com/savvasdalkitsis/gameframe/ip/repository/PreferenceIpRepository.kt
package com.savvasdalkitsis.gameframe.ip.repository

import com.savvasdalkitsis.gameframe.infra.preferences.RxSharedPreferences
import com.savvasdalkitsis.gameframe.ip.model.IpAddress
import com.savvasdalkitsis.gameframe.ip.model.IpBaseHostMissingException
import rx.Observable

class PreferenceIpRepository(private val preferences: RxSharedPreferences) : IpRepository {

    override val ipAddress: Observable<IpAddress>
        get() = preferences.getString(PREF_IP_BASE)
                .map<IpAddress> { IpAddress.parse(it) }
                .flatMap<IpAddress> {
                    if (it.isValid()) {
                        Observable.just(it)
                    } else {
                        Observable.error(IllegalStateException("saved ip address was not valid: $it"))
                    }
                }
                .onErrorResumeNext { Observable.error(IpBaseHostMissingException("Error trying to read ip address from repository", it)) }

    override fun saveIpAddress(ipAddress: IpAddress) {
        preferences.setString(PREF_IP_BASE, ipAddress.toString())
    }

    companion object {
        private val PREF_IP_BASE = "pref_ip_base"
    }

}
